presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/RuleStore.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.SetMultimap;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Set;
+
+public class RuleStore
+{
+    private final ListMultimap<Class<? extends PlanNode>, Rule> rules;
+
+    public RuleStore(ListMultimap<Class<? extends PlanNode>, Rule> rules)
+    {
+        this.rules = rules;
+    }
+
+    public List<Rule> getPossiblyMatchingRules(PlanNode planNode)
+    {
+        return ImmutableList.<Rule>builder()
+                .addAll(rules.get(planNode.getClass()))
```

Comment:
Yes. At least so far. There is a `SetOperation` which is extended by `Union` or `Except`, but then I would expect it to use `Pattern.node(Except.class)` instead of using `SetOperation` directly.